### PR TITLE
Aeson 2.0 compatibility

### PR DESCRIPTION
**Pull request by dysinger**
Add a patch for Aeson 2.0+

### CODE REVIEW

**Review by Vlix — CHANGES_REQUESTED** [review]
Looks good overall. Thanks for the PR!

It looks very backwards compatible, so I'd say this would be a `1.1.2.0` release.

If, next to the requested changes in the inline comments, you could also:
- update the `ChangeLog`, and put it at the top under a new header: `1.1.2.0`
- update the `.cabal` to `1.1.2.0`

We'll see if the CI picks the newest `aeson` or not. (probably not for `stack`)
If not, I'll set a PR for updating the CI first before checking this PR again.

File: safe-json.cabal [hunk 1]
```diff
@@ -62,7 +62,7 @@ library
       src
   default-extensions: OverloadedStrings
   build-depends:
-      aeson >=1.4.1 && <1.6
+      aeson >=1.4.1 && <1.6 || >=2.0.1.0 && <2.1
```
[the remark this answers is not in the document]
Comment (dysinger) [reply safe-json.cabal:65]: IIRC 2.0.1.0 was the first fix version according to the Aeson changelog/upgrade-notes.

[the remark this answers is not in the document]
Comment (dysinger) [reply safe-json.cabal:65]: Here's the link. https://cs-syd.eu/posts/2021-09-11-json-vulnerability I was correct in my previous comment.

**Comment by dysinger** [comment]
I have applied all of your (excellent) recommendations. Thank you.

File: safe-json.cabal (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vlix) [reply safe-json.cabal:65]: I know about the `aeson` fix. (It's whole the reason the major version got bumped to `2`)
I don't see a reason to limit `safe-json`'s dependency, though. If someone (for SOME reason) would want to use `aeson-2.0.0.0`, I don't see a reason why we'd deny it.

Unless `safe-json` doesn't actually compile, I say just keep the dependency as such:
`aeson >= 1.4.1 && < 2.1`

Also, this project uses the `package.yaml` to create the `.cabal`, so please edit that file and build with stack to have it update the `.cabal` automatically.

**Comment by Vlix** [comment]
One more continuing question about the `aeson` dependency limitation and I think you might have missed line 495 and 507 of `src/Data/SafeJSON/Internal.hs` where there are still CPP sections with `HM./KeyMap.` instead of `Map.` (or my view of the file is outdated, but I don't think so?)

**Comment by dysinger** [comment]
Ah. I missed a few. Thanks for your patience.

File: safe-json.cabal (hunk 1 above)
[the remark this answers is not in the document]
Comment (dysinger) [reply safe-json.cabal:65]: I don't think there will ever be a version 2.0.0.0 or 2.0.0.1 but I am happy to oblige.

[the remark this answers is not in the document]
Comment (Vlix) [reply safe-json.cabal:65]: There's very clearly a [version 2.0.0.0](https://hackage.haskell.org/package/aeson-2.0.0.0) on hackage, so I'd say don't make choices for the user of the library. If they want to use anything below `2.0.1.0` that's on them, this library is just supposed to be compatible with `aeson` in general. 👍

**Comment by dysinger** [comment]
I think I caught everything. There is one instance that has HM in it still.

**Comment by dysinger** [comment]
Grr. 8.2.2 is failing.

**Comment by Vlix** [comment]
Yes, the library builds, but the test doesn't.
We've indeed forgotten some parts, it seems: if `aeson-2.0` is used, an `Object` is a `KeyMap`, not a `HashMap`, so we'll have to add an arbitrary instance for that in the `Instances.hs` module and this also reminds me that it might be a good idea to also add a `SafeJSON` instance for `KeyMap`, since now that that's what's used for `Object`s, users will guaranteed run into it.

So if you can make the `safe-json-test` compile and add a `SafeJSON KeyMap` instance, then the CI should pass and we should be ready to deploy a new `safe-json` version. 👍

**Comment by Vlix** [comment]
Btw, I've just merged a PR that updated the CI for (among others) the new GHC 9.0.1, so if you could merge that in, that'd be great.

**Comment by Vlix** [comment]
Is it ok if I continue this one?

I'll merge your PR into a different branch and then make the tests work and add the `KeyMap` instance.

**Pull request merged**